sourcegraph/sourcegraph
# grpc: log non-utf8 string fields in messages

This PR adds more logging in an attempt to help debug #52181. 

The gRPC "internal error" logging interceptor (added in #51749) will now emit the _names_ of any fields (ex: "repo") in any protobuf messages that couldn't be sent (by either clients or servers) because they contained non-utf-8 string fields (which is a [violation of the protobuf spec](https://protobuf.dev/programming-guides/proto3/#scalar)).  

If the `SRC_GRPC_INTERNAL_ERROR_LOGGING_LOG_NON_UTF8_PROTOBUF_MESSAGES_ENABLED` environment variable is set to `true`, the above logs will also contain a snippet of the json representation of the protobuf message for added debug-ability. 
- **Note**: This environment variable is `false` by default (meaning we'll need to manually enable this for sourcegraph.com). It's probably a bad idea to log all non-utf8 protobuf messages by default (since they could contain sensitive information). Customers will need to opt-into this. It should be safe for sourcegraph.com, however as that doesn't contain private code.
- **Note**: Protobuf messages can be quite large, so by default we truncate the message JSON to 1KB before logging it (so that we don't burn a ton of resources). This is tunable via the `SRC_GRPC_INTERNAL_ERROR_LOGGING_LOG_NON_UTF8_PROTOBUF_MESSAGES_MAX_SIZE_BYTES`. (This truncation method is imperfect, as it can lead to "invalid JSON", but it's better than nothing). 

---

Here is an example of what the logs look like:

```
  [       frontend] ERROR gitserver.client.gRPC.internal.error.reporter internalerrs/logging.go:135 grpc: error while marshaling: string field contains invalid UTF-8 {"grpcService": "gitserver.v1.GitserverService", "grpcMethod": "Search", "grpcCode": "Internal", "nonUTF8StringFields": ["repo"], "messageJSON": "{\"repo\":\"github.com/sourcegraph-testing/zap\\ufffdstupid\",\"revisions\":[{}],\"limit\":50000,\"include_diff\":true,\"query\":{\"Value\":{\"DiffMatches\":{\"expr\":\"test\",\"ignore_case\":true}}}}"}
```

---

Other notes:
- This PR adds another utility function in `internal/env` to help get `uint`-formatted environment variables: https://github.com/sourcegraph/sourcegraph/blob/5f8d6e05fc72ccf5302a45b920f20986773387e0/internal/env/env.go#L158-L168
- This PR borrows / adapts a "test" protobuf definition that the official [protobuf-go](https://github.com/protocolbuffers/protobuf-go/blob/v1.30.0/internal/testprotos/news/news.proto) repository uses. This provides a stable protobuf definition that we can write unit tests against (useful for unit testing the reflection logic that discovers the names of the non-utf8 string fields). 

## Test plan

- Unit testing
- Manual testing:
  1. Enable non-utf8 full message logging by setting the following environment variable in your shell:
  ```fish
  SRC_GRPC_INTERNAL_ERROR_LOGGING_LOG_NON_UTF8_PROTOBUF_MESSAGES_ENABLED=true
  ```
  1. Apply the following gitserver diff to force a non-utf8 issue:
  ```patch
  diff --git a/internal/gitserver/protocol/gitserver.go b/internal/gitserver/protocol/gitserver.go
  index 920302c393..bf79ee6eaa 100644
  --- a/internal/gitserver/protocol/gitserver.go
  +++ b/internal/gitserver/protocol/gitserver.go
  @@ -30,7 +30,7 @@ func (r *SearchRequest) ToProto() *proto.SearchRequest {
                  revs = append(revs, rev.ToProto())
          }
          return &proto.SearchRequest{
  -               Repo:                 string(r.Repo),
  +               Repo:                 string(r.Repo) + "\x80stupid",
                  Revisions:            revs,
                  Query:                r.Query.ToProto(),
                  IncludeDiff:          r.IncludeDiff,
  ```
  
  1. Run the following search query on your local instance: `context:global type:diff test`
  1. See an error message similar to the following appear in your terminal
  ```
  [       frontend] ERROR gitserver.client.gRPC.internal.error.reporter internalerrs/logging.go:135 grpc: error while marshaling: string field contains invalid UTF-8 {"grpcService": "gitserver.v1.GitserverService", "grpcMethod": "Search", "grpcCode": "Internal", "nonUTF8StringFields": ["repo"], "messageJSON": "{\"repo\":\"github.com/sourcegraph-testing/zap\\ufffdstupid\",\"revisions\":[{}],\"limit\":50000,\"include_diff\":true,\"query\":{\"Value\":{\"DiffMatches\":{\"expr\":\"test\",\"ignore_case\":true}}}}"}
  ```

  1. You can test truncation by setting `SRC_GRPC_INTERNAL_ERROR_LOGGING_LOG_NON_UTF8_PROTOBUF_MESSAGES_MAX_SIZE_BYTES=20` and re-running the same search query (`context:global type:diff test`), you should see something like the following appear in your terminal:
    ```
   [       frontend] ERROR gitserver.client.gRPC.internal.error.reporter internalerrs/logging.go:135 grpc: error while marshaling: string field contains invalid UTF-8 {"grpcService": "gitserver.v1.GitserverService", "grpcMethod": "Search", "grpcCode": "Internal", "nonUTF8StringFields": ["repo"], "messageJSON": "{\"repo\":\"github.com/...(truncated 167 bytes)"}
    ```
